client.go

```diff
@@ -0,0 +1,449 @@
+		MessageRequest: &reflectionv1.ServerReflectionRequest_FileContainingExtension{
+			FileContainingExtension: &reflectionv1.ExtensionRequest{
+				ContainingType:  string(messageName),
+				ExtensionNumber: int32(extensionNumber),
+			},
+		},
+	})
+}
+
+// AllExtensionNumbers retrieves the tag numbers for all extensions of the given message that
+// are known to the server.
+//
+// This may return a [*connect.Error] indicating the reason the list of extension numbers
+// could not be retrieved (such as "Not Found" if the given message is not known). But if
+// [IsReflectionStreamError] returns true for the returned error, the stream is broken and
+// cannot be used for further operations.
+//
+// This operation sends a request message on the stream and waits for the corresponding
+// response.
+func (cs *ClientStream) AllExtensionNumbers(messageName protoreflect.FullName) ([]protoreflect.FieldNumber, error) {
+	resp, err := cs.send(&reflectionv1.ServerReflectionRequest{
+		MessageRequest: &reflectionv1.ServerReflectionRequest_AllExtensionNumbersOfType{
+			AllExtensionNumbersOfType: string(messageName),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	respExtNumbers := resp.GetAllExtensionNumbersResponse()
+	if respExtNumbers == nil {
+		return nil, errWrongResponseType(resp, "all_extension_numbers")
+	}
+	extNumbers := make([]protoreflect.FieldNumber, len(respExtNumbers.ExtensionNumber))
+	for i, num := range respExtNumbers.ExtensionNumber {
+		extNumbers[i] = protoreflect.FieldNumber(num)
+	}
+	return extNumbers, nil
+}
+
+func (cs *ClientStream) getDescriptors(operation string, req *reflectionv1.ServerReflectionRequest) ([]*descriptorpb.FileDescriptorProto, error) {
+	resp, err := cs.send(req)
+	if err != nil {
+		return nil, err
+	}
+	respDescriptors := resp.GetFileDescriptorResponse()
+	if respDescriptors == nil {
+		return nil, errWrongResponseType(resp, operation)
+	}
+	descriptors := make([]*descriptorpb.FileDescriptorProto, len(respDescriptors.FileDescriptorProto))
+	for i, data := range respDescriptors.FileDescriptorProto {
+		fileDescriptor := &descriptorpb.FileDescriptorProto{}
+		if err := proto.Unmarshal(data, fileDescriptor); err != nil {
+			return nil, fmt.Errorf("reply to %s contained invalid descriptor proto: %w", operation, err)
+		}
+		descriptors[i] = fileDescriptor
+	}
+	return descriptors, nil
+}
+
+func (cs *ClientStream) send(req *reflectionv1.ServerReflectionRequest) (*reflectionv1.ServerReflectionResponse, error) {
+	req.Host = cs.host
+	// Sending on a bidi stream is usually thread-safe. But the replies are in the same order
+	// as the requests. So to prevent concurrent use from interleaving replies (which would
+	// require much more logic here to properly correlate replies with requests), we send and
+	// receive while holding the mutex. This means that this API does not support pipelining
+	// reflection requests (which could in theory reduce latency, but only when the client
```

Comment:
nit: I think we're missing a closing parens after "latency" and before the comma?

Reply:
I've removed the open paren and tried to use hopefully more standard/clearer sentence structure. Better?